File: src/gacha/banner.ts
import { studentContainer } from "../containers/students";
import { GachaPool } from "./pool";

import {
  DEFAULT_BASE_ONE_STAR_RATE,
  DEFAULT_BASE_THREE_STAR_RATE,
  DEFAULT_BASE_TWO_STAR_RATE,
  DEFAULT_EXTRA_RATE,
  DEFAULT_PICKUP_RATE,
  DEFAULT_THREE_STAR_RATE,
} from "./constants";
import { db } from "../db";
import type { Banner, Student, BannerKind } from "../db/client";

export type DetailedGachaBanner = Banner & {
  pickupPoolStudents: Student[];
  extraPoolStudents: Student[];
  additionalThreeStarStudents: Student[];
};

export interface GachaBannerParams {
  id: string;
  name: string;

  date: string;

  threeStarRate?: number;
  pickupRate?: number;
  extraRate?: number;

  pickupPoolStudents?: Student[];
  extraPoolStudents?: Student[];
  additionalThreeStarStudents?: Student[];

  baseOneStarRate?: number;
  baseTwoStarRate?: number;
  baseThreeStarRate?: number;

  kind: BannerKind;
}

class GachaBanner {
  readonly id: string;
  readonly name: string;

  readonly date: Date;
  readonly kind: BannerKind;

  private _threeStarRate: number;
  private _pickupRate: number;
  private _extraRate: number;

  private _pickupPool: GachaPool;
  private _extraPool: GachaPool;
  private _oneStarPool: GachaPool;
  private _twoStarPool: GachaPool;
  private _threeStarPool: GachaPool;

  private _baseOneStarRate: number;
  private _baseTwoStarRate: number;
  private _baseThreeStarRate: number;

  private _isBootstrapped = false;

  constructor(params: GachaBannerParams) {
    this.id = params.id;
    this.name = params.name;

    this.date = new Date(params.date);

    this._threeStarRate = params.threeStarRate ?? DEFAULT_THREE_STAR_RATE;
    this._pickupRate = params.pickupRate ?? DEFAULT_PICKUP_RATE;
    this._extraRate = params.extraRate ?? DEFAULT_EXTRA_RATE;

    this._baseOneStarRate =
      params.baseOneStarRate ?? DEFAULT_BASE_ONE_STAR_RATE;
    this._baseTwoStarRate =
      params.baseTwoStarRate ?? DEFAULT_BASE_TWO_STAR_RATE;
    this._baseThreeStarRate =
      params.baseThreeStarRate ?? DEFAULT_BASE_THREE_STAR_RATE;

    this._pickupPool = new GachaPool(this.pickupRate);
    this._extraPool = new GachaPool(this.extraRate);
    this._oneStarPool = new GachaPool(this.oneStarRate);
    this._twoStarPool = new GachaPool(this.twoStarRate);
    this._threeStarPool = new GachaPool(this.threeStarRate);

    this.kind = params.kind;

    this.populatePools(
      params.pickupPoolStudents,
      params.extraPoolStudents,
      params.additionalThreeStarStudents,
    );
  }

  private isStudentAvailableInRegion(student: Student) {
    switch (this.kind) {
      case "Global":
        return student.isReleasedGlobal;
      case "JP":
        return student.isReleasedJP;
      default:
        return false;
    }
  }

  private isStudentLimited(student: Student) {
    switch (this.kind) {
      case "Global":
        return student.isLimitedGlobal;
      case "JP":
        return student.isLimitedJP;
      default:
        return false;
    }
  }

  private isStudentWelfare(student: Student) {
    switch (this.kind) {
      case "Global":
        return student.isWelfareGlobal;
      case "JP":
        return student.isWelfareJP;
      default:
        return false;
    }
  }

  private isStudentArchive(student: Student) {
    switch (this.kind) {
      case "Global":
        return student.isArchiveGlobal;
      case "JP":
        return student.isArchiveJP;
      default:
        return false;
    }
  }

  private isPullable(
    student: Student,
    additionalCondition: (student: Student) => boolean,
  ): boolean {
    return (
      !this.isStudentLimited(student) &&
      !this.isStudentWelfare(student) &&
      !this.isStudentArchive(student) &&
      this.isStudentAvailableInRegion(student) &&
      additionalCondition(student)
    );
  }

  private isOneStar(student: Student): boolean {
    return student.rarity === 1;
  }

  private isTwoStar(student: Student): boolean {
    return student.rarity === 2;
  }

  private isThreeStar(student: Student): boolean {
    return student.rarity === 3;
  }

  private populatePools(
    pickupPoolStudents?: Student[],
    extraPoolStudents?: Student[],
    additionalThreeStarStudents?: Student[],
  ): void {
    if (this._isBootstrapped) {
      return;
    }

    const oneStarStudents = studentContainer.getStudentsWhere((student) =>
      this.isPullable(student, this.isOneStar),
    );

    const twoStarStudents = studentContainer.getStudentsWhere((student) =>
      this.isPullable(student, this.isTwoStar),
    );

    const threeStarStudents = studentContainer
      .getStudentsWhere((student) => this.isPullable(student, this.isThreeStar))
      .filter(
        (student) =>
          !pickupPoolStudents?.includes(student) &&
          !extraPoolStudents?.includes(student),
      );

    oneStarStudents.forEach((student) => this._oneStarPool.addStudent(student));
    twoStarStudents.forEach((student) => this._twoStarPool.addStudent(student));
    threeStarStudents.forEach((student) =>
      this._threeStarPool.addStudent(student),
    );

    if (pickupPoolStudents) {
      pickupPoolStudents.forEach((student) =>
        this._pickupPool.addStudent(student),
      );
    }

    if (extraPoolStudents) {
      extraPoolStudents.forEach((student) =>
        this._extraPool.addStudent(student),
      );
    }

    if (additionalThreeStarStudents) {
      additionalThreeStarStudents.forEach((student) =>
        this._threeStarPool.addStudent(student),
      );
    }

    this._isBootstrapped = true;
  }

  private calculateRate(base: number): number {
    return (
      base - ((this._threeStarRate - this._baseThreeStarRate) / base) * 100
    );
  }

  get oneStarRate(): number {
    return this.calculateRate(this._baseOneStarRate);
  }

  get twoStarRate(): number {
    return this.calculateRate(this._baseTwoStarRate);
  }

  get threeStarRate(): number {
    return this._threeStarRate - this._pickupRate - this._extraRate;
  }

  get pickupRate(): number {
    return this._pickupRate;
  }

  get extraRate(): number {
    return this._extraRate;
  }

  get pools(): GachaPool[] {
    return [
      this._oneStarPool,
      this._twoStarPool,
      this._threeStarPool,
      this._pickupPool,
      this._extraPool,
    ];
  }

  pullTen(): [Student, string][] {
    let hasAtLeastTwoStar = false;

    const students: [Student, string][] = [];

    for (let i = 0; i < 10; ++i) {
      if (i === 9 && !hasAtLeastTwoStar) {
        const { student } = this.pull(true);

        if (student) {
          students.push(student);
        }
      } else {
        const { student, isOneStar } = this.pull();

        if (!isOneStar) {
          hasAtLeastTwoStar = true;
        }

        if (student) {
          students.push(student);
        }
      }
    }

    if (students.length < 10) {
      throw new Error("Failed to pull 10 students...");
    }

    // ensure last drop is always 2★ or higher
    if (students[students.length - 1][0].rarity === 1) {
      const firstNonOneStar = students.findIndex((student) => {
        return student[0].rarity !== 1;
      });

      if (firstNonOneStar !== -1) {
        const randomTwoStar = students[firstNonOneStar];
        const randomOneStar = students[students.length - 1];

        students[firstNonOneStar] = randomOneStar;
        students[students.length - 1] = randomTwoStar;
      }
    }

    return students;
  }

  pull(ensureNoOneStar = false) {
    const pools = this.pools.slice(0);
    let rates = pools.map((pool) => pool.rate);

    if (ensureNoOneStar) {
      const oneStarIndex = pools.findIndex(
        (pool) => pool === this._oneStarPool,
      );

      if (oneStarIndex !== -1) {
        pools.splice(oneStarIndex, 1);
        rates.splice(oneStarIndex, 1);
      }
    }

    // Normalize rates
    const sumRates = rates.reduce((a, b) => a + b, 0);
    rates = rates.map((rate) => rate / sumRates);

    // Randomly choose a pool based on probabilities
    const poolIndex = this.getRandomIndex(rates);
    const pool = pools[poolIndex];

    const isOneStar = pool === this._oneStarPool;
    const student = pool.pull();

    return {
      student,
      isOneStar,
    };
  }

  isPickup(key: string): boolean {
    return this._pickupPool.hasStudent(key);
  }

  private getRandomIndex(rates: number[]): number {
    const randomValue = Math.random();

    let cumulativeRate = 0;

    for (let i = 0; i < rates.length; ++i) {
      cumulativeRate += rates[i];

      if (randomValue <= cumulativeRate) {
        return i;
      }
    }

    // This should never happen, but who knows.
    return rates.length - 1;
  }

  static fromJSON(params: GachaBannerParams): GachaBanner {
    return new GachaBanner(params);
  }

  static async all() {
    return db.banner
      .findMany({
        orderBy: {
          sortKey: "asc",
        },
        include: {
          pickupPoolStudents: true,
          extraPoolStudents: true,
          additionalThreeStarStudents: true,
        },
      })
      .then((entries) => {
        return entries.map((entry) => GachaBanner.fromDBEntry(entry));
      });
  }

  static fromDBEntry(entry: DetailedGachaBanner): GachaBanner {
    return new GachaBanner({
      id: entry.id,
      name: entry.name,
      date: entry.date,
      threeStarRate: entry.threeStarRate / 1000,
      pickupRate: entry.pickupRate / 1000,
      extraRate: entry.extraRate / 1000,
      pickupPoolStudents: entry.pickupPoolStudents ?? undefined,
      extraPoolStudents: entry.extraPoolStudents ?? undefined,
      additionalThreeStarStudents:
        entry.additionalThreeStarStudents ?? undefined,
      baseOneStarRate: entry.baseOneStarRate / 1000,
      baseTwoStarRate: entry.baseTwoStarRate / 1000,
      baseThreeStarRate: entry.baseThreeStarRate / 1000,
      kind: entry.kind,
    });
  }
}

export { GachaBanner };

File: src/gacha/banner.rs
use super::pool::GachaPool;
use crate::db::enums::BannerKind;
use crate::db::models::{DetailedBanner, Student};
use anyhow::{anyhow, Result};
use rand::Rng;

#[derive(Debug, Clone)]
pub struct GachaBanner {
    pub id: String,
    pub name: String,
    pub kind: BannerKind,
    three_star_rate: f64,
    pickup_rate: f64,
    extra_rate: f64,
    base_one_star_rate: f64,
    base_two_star_rate: f64,
    base_three_star_rate: f64,
    pickup_pool: GachaPool,
    extra_pool: GachaPool,
    one_star_pool: GachaPool,
    two_star_pool: GachaPool,
    three_star_pool: GachaPool,
    additional_three_star_students: Vec<Student>,
}

impl GachaBanner {
    pub fn from_db_entry(entry: DetailedBanner, all_students: &[Student]) -> Self {
        let three_star_rate = entry.banner.three_star_rate as f64 / 1000.0;
        let pickup_rate = entry.banner.pickup_rate as f64 / 1000.0;
        let extra_rate = entry.banner.extra_rate as f64 / 1000.0;
        let base_one_star_rate = entry.banner.base_one_star_rate as f64 / 1000.0;
        let base_two_star_rate = entry.banner.base_two_star_rate as f64 / 1000.0;
        let base_three_star_rate = entry.banner.base_three_star_rate as f64 / 1000.0;

        let mut banner = Self {
            id: entry.banner.id,
            name: entry.banner.name,
            kind: entry.banner.kind,
            three_star_rate,
            pickup_rate,
            extra_rate,
            base_one_star_rate,
            base_two_star_rate,
            base_three_star_rate,
            pickup_pool: GachaPool::new(pickup_rate),
            extra_pool: GachaPool::new(extra_rate),
            one_star_pool: GachaPool::new(0.0),
            two_star_pool: GachaPool::new(0.0),
            three_star_pool: GachaPool::new(0.0),
            additional_three_star_students: entry.additional_three_star_students.clone(),
        };

        banner.one_star_pool.rate = banner.one_star_rate();
        banner.two_star_pool.rate = banner.two_star_rate();
        banner.three_star_pool.rate = banner.three_star_pool_rate();

        banner.populate_pools(
            all_students,
            &entry.pickup_pool_students,
            &entry.extra_pool_students,
            &entry.additional_three_star_students,
        );

        banner
    }

    fn is_student_available(&self, student: &Student) -> bool {
        match self.kind {
            BannerKind::Global => student.is_released_global,
            BannerKind::JP => student.is_released_jp,
        }
    }

    fn is_student_limited(&self, student: &Student) -> bool {
        match self.kind {
            BannerKind::Global => student.is_limited_global,
            BannerKind::JP => student.is_limited_jp,
        }
    }

    fn is_student_welfare(&self, student: &Student) -> bool {
        match self.kind {
            BannerKind::Global => student.is_welfare_global,
            BannerKind::JP => student.is_welfare_jp,
        }
    }

    fn is_student_archive(&self, student: &Student) -> bool {
        match self.kind {
            BannerKind::Global => student.is_archive_global,
            BannerKind::JP => student.is_archive_jp,
        }
    }

    fn is_pullable(&self, student: &Student, rarity: i32) -> bool {
        !self.is_student_limited(student)
            && !self.is_student_welfare(student)
            && !self.is_student_archive(student)
            && self.is_student_available(student)
            && student.rarity == rarity
    }

    fn populate_pools(
        &mut self,
        all_students: &[Student],
        pickup: &[Student],
        extra: &[Student],
        additional: &[Student],
    ) {
        let pickup_ids: Vec<&str> = pickup.iter().map(|s| s.id.as_str()).collect();
        let extra_ids: Vec<&str> = extra.iter().map(|s| s.id.as_str()).collect();

        for student in all_students {
            if self.is_pullable(student, 1) {
                self.one_star_pool.add_student(student.clone());
            } else if self.is_pullable(student, 2) {
                self.two_star_pool.add_student(student.clone());
            } else if self.is_pullable(student, 3)
                && !pickup_ids.contains(&student.id.as_str())
                && !extra_ids.contains(&student.id.as_str())
            {
                self.three_star_pool.add_student(student.clone());
            }
        }

        for student in pickup {
            self.pickup_pool.add_student(student.clone());
        }
        for student in extra {
            self.extra_pool.add_student(student.clone());
        }
        for student in additional {
            self.three_star_pool.add_student(student.clone());
        }
    }

    fn calculate_rate(&self, base: f64) -> f64 {
        base - ((self.three_star_rate - self.base_three_star_rate) / base) * 100.0
    }

    pub fn one_star_rate(&self) -> f64 {
        self.calculate_rate(self.base_one_star_rate)
    }

    pub fn two_star_rate(&self) -> f64 {
        self.calculate_rate(self.base_two_star_rate)
    }

    pub fn three_star_pool_rate(&self) -> f64 {
        self.three_star_rate - self.pickup_rate - self.extra_rate
    }

    pub fn pickup_rate(&self) -> f64 {
        self.pickup_rate
    }

    pub fn extra_rate(&self) -> f64 {
        self.extra_rate
    }

    fn pools(&self) -> [&GachaPool; 5] {
        [
            &self.one_star_pool,
            &self.two_star_pool,
            &self.three_star_pool,
            &self.pickup_pool,
            &self.extra_pool,
        ]
    }

    pub fn pull_ten<R: Rng + ?Sized>(&self, rng: &mut R) -> Result<Vec<(Student, String)>> {
        let mut has_at_least_two_star = false;
        let mut students = Vec::with_capacity(10);

        for i in 0..10 {
            if i == 9 && !has_at_least_two_star {
                if let Some(student) = self.pull(rng, true) {
                    students.push(student);
                }
            } else {
                let result = self.pull(rng, false);
                if let Some((ref student, _)) = result {
                    if student.rarity != 1 {
                        has_at_least_two_star = true;
                    }
                }
                if let Some(student) = result {
                    students.push(student);
                }
            }
        }

        if students.len() < 10 {
            return Err(anyhow!("Failed to pull 10 students..."));
        }

        // Ensure last drop is always 2★ or higher
        if students.last().map(|(s, _)| s.rarity) == Some(1) {
            if let Some(first_non_one) = students.iter().position(|(s, _)| s.rarity != 1) {
                let last = students.len() - 1;
                students.swap(first_non_one, last);
            }
        }

        Ok(students)
    }

    pub fn pull<R: Rng + ?Sized>(
        &self,
        rng: &mut R,
        ensure_no_one_star: bool,
    ) -> Option<(Student, String)> {
        let mut pools: Vec<&GachaPool> = self.pools().to_vec();
        let mut rates: Vec<f64> = pools.iter().map(|p| p.rate).collect();

        if ensure_no_one_star {
            if let Some(idx) = pools.iter().position(|p| std::ptr::eq(*p, &self.one_star_pool)) {
                pools.remove(idx);
                rates.remove(idx);
            }
        }

        let sum: f64 = rates.iter().sum();
        if sum <= 0.0 {
            return None;
        }
        rates.iter_mut().for_each(|r| *r /= sum);

        let pool_index = get_random_index(rng, &rates);
        let pool = pools.get(pool_index)?;
        pool.pull(rng)
    }

    pub fn is_pickup(&self, key: &str) -> bool {
        self.pickup_pool.has_student(key)
    }

    pub fn is_extra(&self, key: &str) -> bool {
        self.extra_pool.has_student(key)
    }

    pub fn is_additional_three_star(&self, key: &str) -> bool {
        self.additional_three_star_students
            .iter()
            .any(|s| s.id == key)
    }
}

fn get_random_index<R: Rng + ?Sized>(rng: &mut R, rates: &[f64]) -> usize {
    let random_value: f64 = rng.gen();
    let mut cumulative = 0.0;
    for (i, rate) in rates.iter().enumerate() {
        cumulative += rate;
        if random_value <= cumulative {
            return i;
        }
    }
    rates.len().saturating_sub(1)
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand_chacha::ChaCha8Rng;

    fn dummy_student(id: &str, rarity: i32) -> Student {
        use crate::db::enums::*;
        Student {
            id: id.to_string(),
            dev_name: id.to_string(),
            schale_db_id: 1,
            default_order: 0,
            name: id.to_string(),
            last_name: id.to_string(),
            first_name: id.to_string(),
            name_jp: String::new(),
            last_name_jp: String::new(),
            first_name_jp: String::new(),
            school: School::Abydos,
            club: Club::EmptyClub,
            age: "16".into(),
            birthday: "January 1".into(),
            introduction: None,
            hobbies: None,
            voice_actor: None,
            illustrator: None,
            designer: None,
            height: None,
            memorobi_level: 0,
            combat_class: CombatClass::Main,
            combat_role: CombatRole::DamageDealer,
            combat_position: CombatPosition::Front,
            attack_type: AttackType::Explosion,
            defense_type: DefenseType::LightArmor,
            street_battle_adaptation: 0,
            outdoor_battle_adaptation: 0,
            indoor_battle_adaptation: 0,
            weapon_type: WeaponType::AR,
            uses_cover: true,
            equipment: vec![],
            has_bond_gear_jp: false,
            has_bond_gear_global: false,
            has_bond_gear_cn: false,
            rarity,
            is_released_jp: true,
            is_released_global: true,
            is_released_cn: false,
            is_welfare_jp: false,
            is_welfare_global: false,
            is_welfare_cn: false,
            is_limited_jp: false,
            is_limited_global: false,
            is_limited_cn: false,
            is_fest_jp: false,
            is_fest_global: false,
            is_fest_cn: false,
            is_archive_jp: false,
            is_archive_global: false,
            is_archive_cn: false,
            search_tags: vec![],
            base_variant_id: None,
        }
    }

    #[test]
    fn pull_ten_always_returns_ten_and_last_not_one_star_when_possible() {
        let mut ones = Vec::new();
        let mut twos = Vec::new();
        let mut threes = Vec::new();
        for i in 0..20 {
            ones.push(dummy_student(&format!("1_{i}"), 1));
            twos.push(dummy_student(&format!("2_{i}"), 2));
            threes.push(dummy_student(&format!("3_{i}"), 3));
        }
        let all: Vec<_> = ones
            .iter()
            .chain(twos.iter())
            .chain(threes.iter())
            .cloned()
            .collect();

        let entry = DetailedBanner {
            banner: crate::db::models::BannerRow {
                id: "regular".into(),
                name: "Regular".into(),
                sort_key: 1,
                date: "2024-01-01".into(),
                three_star_rate: 30,
                pickup_rate: 0,
                extra_rate: 0,
                base_one_star_rate: 785,
                base_two_star_rate: 185,
                base_three_star_rate: 30,
                kind: BannerKind::Global,
            },
            pickup_pool_students: vec![],
            extra_pool_students: vec![],
            additional_three_star_students: vec![],
        };

        let banner = GachaBanner::from_db_entry(entry, &all);
        let mut rng = ChaCha8Rng::seed_from_u64(42);

        for _ in 0..50 {
            let pulls = banner.pull_ten(&mut rng).unwrap();
            assert_eq!(pulls.len(), 10);
            assert!(pulls.last().unwrap().0.rarity >= 2);
        }
    }
}
